File: backend/src/main/java/com/erp/cuahangtienloi/entity/BangLuong.java
package com.erp.cuahangtienloi.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "bang_luong")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class BangLuong {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "id_nhan_vien")
    private UUID idNhanVien;

    @Column(name = "id_chi_nhanh")
    private UUID idChiNhanh;

    @Column(name = "loai_hop_dong")
    private String loaiHopDong;

    @Column(name = "thang_nam")
    private String thangNam;

    @Column(name = "tong_gio_lam")
    private BigDecimal tongGioLam;

    @Column(name = "overtime_hours")
    private BigDecimal overtimeHours;

    @Column(name = "tong_so_ca")
    private Integer tongSoCa;

    @Column(name = "gio_dieu_chinh")
    private BigDecimal gioDieuChinh;

    @Column(name = "ly_do_dieu_chinh", columnDefinition = "TEXT")
    private String lyDoDieuChinh;

    @Column(name = "luong_theo_gio")
    private BigDecimal luongTheoGio;

    @Column(name = "luong_cung")
    private BigDecimal luongCung;

    @Column(name = "luong_cung_thuc_te")
    private BigDecimal luongCungThucTe;

    @Column(name = "tien_cong_theo_gio")
    private BigDecimal tienCongTheoGio;

    @Column(name = "tien_ot")
    private BigDecimal tienOt;

    @Column(name = "thuong")
    private BigDecimal thuong;

    @Column(name = "khau_tru")
    private BigDecimal khauTru;

    @Column(name = "tong_tien_luong")
    private BigDecimal tongTienLuong;

    @Column(name = "trang_thai")
    private String trangThai;

    @Column(name = "id_nguoi_xac_nhan")
    private UUID idNguoiXacNhan;

    @Column(name = "ngay_xac_nhan")
    private LocalDateTime ngayXacNhan;

    @Column(name = "id_nguoi_duyet_chi")
    private UUID idNguoiDuyetChi;

    @Column(name = "ngay_duyet_chi")
    private LocalDateTime ngayDuyetChi;

    @Column(name = "id_nguoi_thanh_toan")
    private UUID idNguoiThanhToan;

    @Column(name = "ngay_thanh_toan")
    private LocalDateTime ngayThanhToan;

    @Column(name = "ma_phieu_chi")
    private String maPhieuChi;

    @Column(name = "ngay_tao")
    private LocalDateTime ngayTao;

    @Column(name = "ngay_cap_nhat")
    private LocalDateTime ngayCapNhat;
}

File: frontend/src/api/bangLuong.ts
import { API_BASE_URL } from '@/config/api';
import { getAuthHeaders } from './http';

export interface BangLuongDTO {
  id: string;
  idNhanVien: string;
  tenNhanVien?: string;
  maNhanVien?: string;
  vaiTro?: string;
  idChiNhanh: string;
  tenChiNhanh?: string;
  loaiHopDong: string;
  thangNam: string;
  tongGioLam: number;
  overtimeHours: number;
  tongSoCa: number;
  gioDieuChinh?: number | null;
  lyDoDieuChinh?: string;
  luongTheoGio: number;
  luongCung: number;
  luongCungThucTe: number;
  tienCongTheoGio: number;
  tienOt: number;
  thuong: number;
  khauTru: number;
  tongTienLuong: number;
  trangThai: string;
  idNguoiXacNhan?: string;
  tenNguoiXacNhan?: string;
  ngayXacNhan?: string;
  idNguoiDuyetChi?: string;
  tenNguoiDuyetChi?: string;
  ngayDuyetChi?: string;
  idNguoiThanhToan?: string;
  tenNguoiThanhToan?: string;
  ngayThanhToan?: string;
  maPhieuChi?: string;
}

const getHeaders = (): HeadersInit => {
  return getAuthHeaders();
};

export const bangLuongApi = {
  getAll: async (period?: string): Promise<BangLuongDTO[]> => {
    const query = period ? `?period=${encodeURIComponent(period)}` : '';
    const response = await fetch(`${API_BASE_URL}/api/bang-luong${query}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương');
    return response.json();
  },

  getMine: async (period?: string): Promise<BangLuongDTO[]> => {
    const query = period ? `?period=${encodeURIComponent(period)}` : '';
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/me${query}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương của bạn');
    return response.json();
  },

  getById: async (id: string): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương');
    return response.json();
  },

  getByNhanVien: async (idNhanVien: string): Promise<BangLuongDTO[]> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/by-employee/${idNhanVien}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương');
    return response.json();
  },

  getByThangNam: async (thangNam: string): Promise<BangLuongDTO[]> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/by-month/${thangNam}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương');
    return response.json();
  },

  getByStatus: async (trangThai: string): Promise<BangLuongDTO[]> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/by-status/${trangThai}`, {
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to fetch bảng lương');
    return response.json();
  },

  /** Tự tổng hợp bảng lương 1 tháng từ dữ liệu chấm công. */
  generate: async (thangNam: string): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/generate/${thangNam}`, {
      method: 'POST',
      headers: getHeaders(),
    });
    if (!response.ok) {
      const err = await response.json().catch(() => null);
      throw new Error(err?.message || 'Failed to generate bảng lương');
    }
  },

  create: async (data: BangLuongDTO): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...getHeaders() },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to create');
    return response.json();
  },

  update: async (id: string, data: Partial<BangLuongDTO>): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json', ...getHeaders() },
      body: JSON.stringify(data),
    });
    if (!response.ok) throw new Error('Failed to update');
    return response.json();
  },

  adjustHours: async (id: string, hours: number, reason: string): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}/hours-adjustment`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json', ...getHeaders() },
      body: JSON.stringify({ hours, reason }),
    });
    if (!response.ok) {
      throw new Error((await response.json().catch(() => null))?.message || 'Không thể lưu điều chỉnh giờ');
    }
    return response.json();
  },

  confirmHours: async (id: string): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}/confirm-hours`, {
      method: 'POST', headers: getHeaders(),
    });
    if (!response.ok) throw new Error((await response.json().catch(() => null))?.message || 'Không thể xác nhận giờ làm');
    return response.json();
  },

  approvePayment: async (id: string): Promise<BangLuongDTO> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}/approve-payment`, {
      method: 'POST', headers: getHeaders(),
    });
    if (!response.ok) throw new Error((await response.json().catch(() => null))?.message || 'Không thể duyệt chi lương');
    return response.json();
  },

  approvePaymentBatch: async (ids: string[]): Promise<BangLuongDTO[]> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/approve-payment/batch`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...getHeaders() },
      body: JSON.stringify({ ids }),
    });
    if (!response.ok) throw new Error((await response.json().catch(() => null))?.message || 'Không thể duyệt chi các bảng lương đã chọn');
    return response.json();
  },

  delete: async (id: string): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/api/bang-luong/${id}`, {
      method: 'DELETE',
      headers: getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to delete');
  },
};
